File: week11/2/Prob3_prim.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
   int mi_node(int n, int dist[], bool vis[]) { 
      int mi = INT_MAX; 
      int mi_idx = -1; 
      for (int i = 0; i < n; i++) { 
        if (vis[i] == false && dist[i] < mi) { 
          mi = dist[i], mi_idx = i; 
        } 
      } 
      return mi_idx; 
    } 

    int prim(vector<vector<int>> &edges, int n) {
        int parent[n]; 
        int dist[n]; 
        bool vis[n]; 
        for (int i = 0; i < n; i++) { 
            dist[i] = numeric_limits<int>::max(); 
            vis[i] = false; 
        } 
        parent[0] = -1; 
        dist[0] = 0; 
        for (int i = 0; i < n - 1; i++) { 
            int u = mi_node(n, dist, vis); 
            vis[u] = true; 
            for (int v = 0; v < n; v++) { 
              if (edges[u][v] && vis[v] == false &&  edges[u][v] < dist[v]) { 
                dist[v] = edges[u][v]; 
                parent[v] = u; 
              } 
            } 
          } 
          int cost = 0; 
          for (int i = 1; i < n; i++) {
              cost += edges[parent[i]][i];
          }
        return cost;
    }

    int minCostConnectPoints(vector<vector<int>>& points) {
        int n = points.size(); 
        vector<vector<int>> edges(n, vector<int>(n));
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                int w = abs(points[i][0] - points[j][0]) + abs(points[i][1] - points[j][1]);
                edges[i][j] = w;
            }
        }
        return prim(edges, n);
    }
};

File: week11/2/Prob3_prim.py
from typing import List
from heapq import heappop, heappush

class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        # applying prim's algorithm
        # initialize edge node list
        N = len(points)
        adj = {i:[] for i in range(N)}
        for i in range(N):
            x1, y1 = points[i]
            for j in range(i+1, N):
                x2, y2 = points[j]
                dist = abs(x1-x2) + abs(y1-y2)
                adj[i].append((dist, j))
                adj[j].append((dist, i))
                
        ans = 0
        visited = set()
        cost_idx = [(0, 0)]
        # MST에서 edge의 길이는 N-1개
        while len(visited) < N:
            cost, idx = heappop(cost_idx)
            if idx in visited:
                continue
                
            visited.add(idx)
            ans += cost
            
            for nei in adj[idx]:
                if nei[1] not in visited:
                    heappush(cost_idx, nei)
                    
        return ans
    
sol = Solution()

points = [[2,-3],[-17,-8],[13,8],[-17,-15]]

print(sol.minCostConnectPoints(points))
